File: 16-04-2024/Minimize the Difference.js
//{ Driver Code Starts

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.trim().split("\n").map(string => {
        return string.trim();
    });

    main();
});

function readLine() {
    return inputString[currentLine++];
}


function main() {
    let t = parseInt(readLine());
    let i = 0;
    for(;i<t;i++){
        
        let n = parseInt(readLine());
        
        
        let k = parseInt(readLine());
        
        
        let arr = new Array(n);
        let input_arr = readLine().split(' ').map(x=>parseInt(x));
        for(let i=0;i<n;i++){
            arr[i] = input_arr[i];
        }
        
        let obj = new Solution();
        let res = obj.minimizeDifference(n, k, arr);
        console.log(res);
    }
}

// } Driver Code Ends


//Your Code Start
class Solution {
    
    minimizeDifference(length, k, array) {
        let postMax = new Array(length);
        let postMin = new Array(length);

        postMin[length - 1] = array[length - 1];
        postMax[length - 1] = array[length - 1];

        for (let i = length - 2; i >= 0; --i) {
            postMax[i] = Math.max(array[i], postMax[i + 1]);
            postMin[i] = Math.min(array[i], postMin[i + 1]);
        }

        let minDiff = postMax[k] - postMin[k];

        let pMin = array[0];
        let pMax = array[0];

        for (let i = 1; i < length - k; ++i) {
            let currentMin = Math.max(pMax, postMax[i + k]) - Math.min(pMin, postMin[i + k]);
            minDiff = Math.min(minDiff, currentMin);
            pMax = Math.max(array[i], pMax);
            pMin = Math.min(array[i], pMin);
        }

        minDiff = Math.min(minDiff, pMax - pMin);

        return minDiff;
    }
}
//Your Code End
        

File: 16-04-2024/Minimize the Difference.py
from typing import List


#Your Code Start
class Solution:
    
    # Function to minimize the difference between maximum and minimum element in subarrays of length k.
    def minimizeDifference(self, length: int, k: int, array: List[int]) -> int:
        
        # List to store the maximum element of subarrays starting from i to end.
        post_max = [0] * length
        
        # List to store the minimum element of subarrays starting from i to end.
        post_min = [0] * length

        # Initializing the last elements of post_max and post_min.
        post_min[-1] = array[-1]
        post_max[-1] = array[-1]

        # Iterating backwards to calculate post_max and post_min.
        for i in range(length - 2, -1, -1):
            post_max[i] = max(array[i], post_max[i + 1])
            post_min[i] = min(array[i], post_min[i + 1])

        # Calculating initial difference by removing the first subarray of length k.
        min_diff = post_max[k] - post_min[k]

        # Initializing variables for storing current min and max values.
        p_min = array[0]
        p_max = array[0]

        # Iterating to find the minimum difference.
        for i in range(1, length - k):
            curr_min = max(p_max, post_max[i + k]) - min(p_min, post_min[i + k])
            min_diff = min(min_diff, curr_min)
            p_max = max(array[i], p_max)
            p_min = min(array[i], p_min)
        
        # Calculating difference by removing the last subarray of length k.
        min_diff = min(min_diff, p_max - p_min)

        # Returning the minimum difference.
        return min_diff     
#Your Code End


#{ 
 # Driver Code Starts

class IntArray:
    def __init__(self) -> None:
        pass
    def Input(self,n):
        arr=[int(i) for i in input().strip().split()]#array input
        return arr
    def Print(self,arr):
        for i in arr:
            print(i,end=" ")
        print()


if __name__=="__main__":
    t = int(input())
    for _ in range(t):
        
        n = int(input())
        
        
        k = int(input())
        
        
        arr=IntArray().Input(n)
        
        obj = Solution()
        res = obj.minimizeDifference(n, k, arr)
        
        print(res)
        

# } Driver Code Ends
